src/libopensc/card.c

```diff
@@ -659,15 +659,20 @@ int sc_read_binary(sc_card_t *card, unsigned int idx,
 		r = card->ops->read_binary(card, idx, buf, chunk, flags);
 		if (r == 0 || r == SC_ERROR_FILE_END_REACHED)
 			break;
-		if ((idx > SIZE_MAX - (size_t) r)
-				|| (size_t) r > todo) {
-			/* `idx + r` or `todo - r` would overflow */
-			r = SC_ERROR_OFFSET_TOO_LARGE;
+		if (r < 0 && todo != count) {
+			/* the last command failed, but previous ones succeeded.
+			 * Let's just return what we've successfully read. */
```

Comment:
I would propose to convert this comment into a debug log as it might be unexpected that after a read fails, this function would return success with some bytes.